galley/pkg/config/processor/transforms/ingress/gateway.go

```diff
@@ -0,0 +1,170 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain ingressAdapter copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingress
+
+import (
+	"fmt"
+	"path"
+
+	ingress "k8s.io/api/extensions/v1beta1"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/event"
+	"istio.io/istio/galley/pkg/config/processing"
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/config/processor/transforms/ingress/annotations"
+	"istio.io/istio/galley/pkg/config/resource"
+	"istio.io/istio/galley/pkg/config/synthesize"
+)
+
+type gatewayXform struct {
+	*event.FnTransform
+
+	options processing.ProcessorOptions
+}
+
+var _ event.Transformer = &gatewayXform{}
+
+func newGatewayXform(o processing.ProcessorOptions) event.Transformer {
+	xform := &gatewayXform{}
+	xform.FnTransform = event.NewFnTransform(
+		collection.Names{metadata.K8SExtensionsV1Beta1Ingresses},
+		collection.Names{metadata.IstioNetworkingV1Alpha3Gateways},
+		nil, nil,
+		xform.handle)
+	xform.options = o
+
+	return xform
+}
+
+func (g *gatewayXform) handle(e event.Event, h event.Handler) {
+
+	if g.options.MeshConfig.IngressControllerMode == meshconfig.MeshConfig_OFF {
+		// short circuit and return
+		return
+	}
+
+	switch e.Kind {
+	case event.Added, event.Updated:
+		if !shouldProcessIngress(g.options.MeshConfig, e.Entry) {
+			return
+		}
+
+		gw := g.convertIngressToGateway(e.Entry)
+		evt := event.Event{
+			Kind:   e.Kind,
+			Source: metadata.IstioNetworkingV1Alpha3Gateways,
+			Entry:  gw,
+		}
+		h.Handle(evt)
+
+	case event.Deleted:
+		gw := g.convertIngressToGateway(e.Entry)
+		evt := event.Event{
+			Kind:   e.Kind,
+			Source: metadata.IstioNetworkingV1Alpha3Gateways,
+			Entry:  gw,
+		}
+		evt.Entry.Metadata.Name = generateSyntheticGatewayName(e.Entry.Metadata.Name)
+		evt.Entry.Metadata.Version = generateSyntheticVersion(e.Entry.Metadata.Version)
+
+		h.Handle(evt)
+
+	default:
+		scope.Errorf("gatewayXform.handle: unknown event: %v", e)
+	}
+}
+
+func (g *gatewayXform) convertIngressToGateway(e *resource.Entry) *resource.Entry {
+	namespace, name := e.Metadata.Name.InterpretAsNamespaceAndName()
+
+	var gateway *v1alpha3.Gateway
+	if e.Item != nil {
+		i := e.Item.(*ingress.IngressSpec)
+
+		gateway = &v1alpha3.Gateway{
+			Selector: IstioIngressWorkloadLabels,
+		}
+
+		// FIXME this is ingressAdapter temporary hack until all test templates are updated
```

Comment:
Does this need to be addressed before checkin? If not, maybe switch to the more standard TODO?